File: algorithms/dynamic-programming/sos-dp/go/sos_dp.go
package main

import (
	"fmt"
	"strings"
)

func sosDp(n int, f []int) []int {
	size := 1 << n
	sos := make([]int, size)
	copy(sos, f)

	for i := 0; i < n; i++ {
		for mask := 0; mask < size; mask++ {
			if mask&(1<<i) != 0 {
				sos[mask] += sos[mask^(1<<i)]
			}
		}
	}
	return sos
}

func main() {
	var n int
	fmt.Scan(&n)
	size := 1 << n
	f := make([]int, size)
	for i := 0; i < size; i++ {
		fmt.Scan(&f[i])
	}
	result := sosDp(n, f)
	strs := make([]string, size)
	for i, v := range result {
		strs[i] = fmt.Sprintf("%d", v)
	}
	fmt.Println(strings.Join(strs, " "))
}

File: algorithms/dynamic-programming/sos-dp/cpp/sos_dp.cpp
#include <iostream>
#include <vector>
using namespace std;

vector<int> sosDp(int n, vector<int>& f) {
    int size = 1 << n;
    vector<int> sos(f.begin(), f.end());

    for (int i = 0; i < n; i++) {
        for (int mask = 0; mask < size; mask++) {
            if (mask & (1 << i)) {
                sos[mask] += sos[mask ^ (1 << i)];
            }
        }
    }
    return sos;
}

int main() {
    int n;
    cin >> n;
    int size = 1 << n;
    vector<int> f(size);
    for (int i = 0; i < size; i++) cin >> f[i];
    vector<int> result = sosDp(n, f);
    for (int i = 0; i < size; i++) {
        if (i > 0) cout << ' ';
        cout << result[i];
    }
    cout << endl;
    return 0;
}
